[libc] Fix malloc Block alignment issue on riscv32

mysterymath opened the pull request:
Aligning blocks to max_align_t is neither necessary nor sufficient to ensure that the usable_space() is so aligned. Instead, we make this an invariant of Block and maintain it in init() and split().

This allows targets like riscv32 with small pointers and large max_align_t to maintain the property that all available blocks are aligned for malloc(). This change adjusts the tests to match and also updates them closer to llvm-libc style.

mysterymath committed:
[libc] Fix malloc Block alignment issue on riscv32

Aligning blocks to max_align_t is neither necessary nor sufficient to
ensure that the usable_space() is so aligned. Instead, we make this an
invariant of Block and maintain it in init() and split().

This allows targets like riscv32 with small pointers and large
max_align_t to maintain the property that all available blocks are
aligned for malloc().

mysterymath commented:
This patch had a bit too much [NFC] stuff folded into it, so I stripped out that out to hopefully make it easier to review. PTAL!

PiJoules reviewed (commented):
For each of the comments that mention returning something aligned to `max_align_t`, could you add a `LIBC_ASSERT` for the return value to ensure this?

PiJoules reviewed (commented):
Kind of a large nit, but in a separate NFC patch, could you also introduce something like `typdef size_t prev_field_t` and replace all the hardcoded `sizeof(size_t)`s with `sizeof(prev_field_t)`? This way it's a little more informative on what the individual `sizeof(size_t)`s are which I assume pertain to the `prev_`field.

mysterymath commented:
> Kind of a large nit, but in a separate NFC patch, could you also introduce something like `typdef size_t prev_field_t` and replace all the hardcoded `sizeof(size_t)`s with `sizeof(prev_field_t)`? This way it's a little more informative on what the individual `sizeof(size_t)`s are which I assume pertain to the `prev_`field.

Yeah, I think that's a good idea, although I'd probably break out a `Block::prev_field_size` global rather than making a new type. I'll add it to my NFC patch chain that will follow this one.

PiJoules reviewed (commented):
Are  `ALIGNMENT` still needed after this patch? It looks like we unconditionally align up to multiples of `alignof(max_align_t)`.

mysterymath commented on libc/src/__support/block.h:
```diff
@@ -129,8 +130,9 @@ class Block {
 
-  /// Creates the first block for a given memory region, followed by a sentinel
-  /// last block. Returns the first block.
+  /// Initializes a given memory region into a first block and a sentinel last
+  /// block. Returns the first block, which has its usable space aligned to
+  /// max_align_t.
```
This is an invariant for all Blocks, so I've checked for this in the block constructor. There's no way for either the Block or usable space pointer to change at runtime, so this should be sufficient to globally maintain the invariant.

mysterymath commented on libc/src/__support/block.h:
```diff
@@ -186,6 +188,9 @@ class Block {
 
   /// @returns A pointer to the usable space inside this block.
+  ///
+  /// Unless specifically requested otherwise, this will be aligned to
+  /// max_align_t.
```
See above; establishing these in the constructor should allow us to avoid repeating the checks throughout the implementation.

mysterymath replied:
Nvm; see below. Added.

mysterymath commented:
> Are `ALIGNMENT` still needed after this patch? It looks like we unconditionally align up to multiples of `alignof(max_align_t)`.

It's not, but removing it was one of the things that muddied this patch. I'll go through and strip it out in a follow-up NFC patch.

mysterymath commented:
> For each of the comments that mention returning something aligned to `max_align_t`, could you add a `LIBC_ASSERT` for the return value to ensure this?

I've added this to the constructor and mutators, since there are fewer of these than places that expose the alignment invariant, and those are the only places it could break. I had missed an assertion in `::split`.

mysterymath committed:
Add more usable space alignment assertions

mysterymath commented:
> > For each of the comments that mention returning something aligned to `max_align_t`, could you add a `LIBC_ASSERT` for the return value to ensure this?
> 
> I've added this to the constructor and mutators, since there are fewer of these than places that expose the alignment invariant, and those are the only places it could break. I had missed an assertion in `::split`.

Actually now that I think of it, the fact that this was missed means that it's probably a good idea to check in the accessors too. I'll add that.

PiJoules commented on libc/src/__support/block.h:
```diff
@@ -248,46 +259,48 @@ class Block {
-    uintptr_t next_usable_space = align_up(start + BLOCK_OVERHEAD, alignment);
-    uintptr_t next_block = next_usable_space - BLOCK_OVERHEAD;
-    return next_block - start + sizeof(prev_);
+    // The minimum advancment is sizeof(Block), since the resulting position may
+    // happen to be aligned. What is the maximum? Advancing by sizeof(Block)
+    // leaves the result still aligned to alignof(Block). So the most additional
+    // advancement required would be if the point is exactly alignof(Block) past
+    // alignment. The remaining size to the next alignment would then be
+    // alignment - alignof(Block). So the total maximum advancement required is
+    // sizeof(Block) + alignment - alignof(Block).
```
Had some trouble thinking about why the max is calculated this way but @mysterymath explained offline and I'm going to write down the explanation which can hopefully be added here to expand on the max calculation.

```
When picking a new size, we need to ensure we can always pick a point inside that block that is aligned to alignment. We also need to make sure that advancing by the requested size will ensure the address at the end of the new size is Block-aligned (so we can always allocate new Blocks after it).

Assuming the alignment is just one max_align_t, we can always return the size size the new inner size would always be aligned. Otherwise, we need to advance by a little extra space.

The minimum additional advancement is sizeof(Block), since the resulting position may happen to be aligned. This space is also needed just to hold the next block this can be partitioned into.

What is the maximum? We know that requesting an additional size by `alignment` will always ensure there's a point in the new allocation that will be aligned to `alignment`, so we can add that to our new size. However, let's say our new usable_space would already be aligned to `alignment`, then advancing by `alignment` would not be necessary since the usable_space would already be aligned. If instead the usable_space were one byte after an alignment point, then we'd only need to advance the size by at most `alignment - 1` byte to find a point in the new block that can be aligned to `alignment`. Over the contiguous memory region Blocks operate under, any Block split always ensures partitions are at minimum Block-aligned. Therefore, the most we need to advance is actually `alignment - alignof(Block)`. This way, even if advancing by `size + sizeof(Block)` happens to be aligned to `alignment`, we don't need to request the full `alignment`. Leaving out the `alignof(Block)` also ensures the partition after this one will always be Block-aligned.
```

Feel free to correct/expand on anything I got wrong with my understanding here.

mysterymath replied:
Thanks, I've rewritten this to be more explicit about the goals and mechanisms in light of our discussion.

PiJoules commented on libc/src/__support/block.h:
```diff
@@ -309,21 +322,30 @@ class Block {
 
+  LIBC_INLINE static uintptr_t next_possible_block_start(
+      uintptr_t ptr, size_t usable_space_alignment = alignof(max_align_t)) {
+    return align_up(ptr + sizeof(Block), usable_space_alignment) -
```
Feel free to disregard this since this is probably me nitpicking on potential overflows.

mysterymath replied:
I do think this code should be made resistant to overflows induced by user behavior at its exterior surface (`_end`, `__llvm_libc_heap_limit`, and the values passed to allocation functions). I definitely haven't been all that careful with this across the board, so I've filed #122978 for this.

mysterymath replied:
These conveniently return the correct answer in modulo arithmetic, so I modified their uses to detect overflow and react accordingly.

mysterymath committed:
Address overflow cases in init

mysterymath committed:
Handle overflow cases in split

mysterymath committed:
Revert change away from kN

mysterymath commented on libc/test/src/__support/block_test.cpp:
```diff
@@ -355,93 +346,84 @@ TEST(LlvmLibcBlockTest, CanGetConstBlockFromUsableSpace) {
   // Ensure we can allocate everything up to the block size within this block.
-  for (size_t i = 0; i < kN - 2 * Block::BLOCK_OVERHEAD; ++i) {
-    alignas(Block::ALIGNMENT) array<byte, kN> bytes{};
+  for (size_t i = 0; i < 1024; ++i) {
```
Done.

mysterymath committed:
Clarify max size calculaton comment

nickdesaulniers commented on libc/test/src/__support/freelist_heap_test.cpp:
```diff
@@ -241,16 +242,15 @@ TEST_FOR_EACH_ALLOCATOR(AlignedAlloc, 2048) {
+// still get aligned allocations even if the underlying buffer is not aligned to
+// the alignments we request.
+TEST(LlvmLibcFreeListHeap, AlignedAllocUnalignedBuffer) {
   constexpr size_t BUFFER_SIZE = 4096;
```
```
llvm-project/libc/test/src/__support/freelist_heap_test.cpp:249:20: warning: unused variable 'BUFFER_SIZE' [-Wunused-variable]
  249 |   constexpr size_t BUFFER_SIZE = 4096;
      |                    ^~~~~~~~~~~
```

nickdesaulniers commented on libc/src/__support/block.h:
```diff
@@ -186,11 +174,19 @@ class Block {
-    return reinterpret_cast<const cpp::byte *>(this) + BLOCK_OVERHEAD;
+    const auto *s = reinterpret_cast<const cpp::byte *>(this) + BLOCK_OVERHEAD;
+    LIBC_ASSERT(reinterpret_cast<uintptr_t>(s) % alignof(max_align_t) == 0 &&
+                "usable space must be aligned to a multiple of max_align_t");
```
```
llvm-project/libc/src/__support/block.h:188: Assertion failed: 'reinterpret_cast<uintptr_t>(s) % alignof(max_align_t) == 0 && "usable space must be aligned to a multiple of max_align_t"' in function: 'const cpp::byte *__llvm_libc_20_0_0_git::Block::usable_space() const'
```

nickdesaulniers reviewed (commented):
Hi @mysterymath , the 32b riscv buidbot is failing at this PR with the two errors. Can you PTAL?

mysterymath commented:
> Hi @mysterymath , the 32b riscv buidbot is failing at this PR with the two errors. Can you PTAL?

Sorry about this; I was able to fix forward:

https://lab.llvm.org/buildbot/#/builders/196/builds/3750

4-byte pointers and 16-byte max_align_t has been a stress-test for Block; I'm both grateful and vaguely horrified at such a large alignment on such a small platform.

nickdesaulniers commented:
> Sorry about this; I was able to fix forward:

Thx!
